File: library-core/src/main/java/com/mikepenz/fastadapter/adapters/ItemAdapter.java
package com.mikepenz.fastadapter.adapters;

import com.mikepenz.fastadapter.IInterceptor;
import com.mikepenz.fastadapter.IItem;
import com.mikepenz.fastadapter.IItemList;

import androidx.recyclerview.widget.RecyclerView;

/**
 * Created by mikepenz on 27.12.15.
 * A general ItemAdapter implementation based on the AbstractAdapter to speed up development for general items
 */
public class ItemAdapter<Item extends IItem<? extends RecyclerView.ViewHolder>> extends ModelAdapter<Item, Item> {

    public ItemAdapter() {
        super((IInterceptor<Item, Item>) IInterceptor.Companion.getDEFAULT());
    }

    public ItemAdapter(IItemList<Item> itemList) {
        super(itemList, (IInterceptor<Item, Item>) IInterceptor.Companion.getDEFAULT());
    }

    /**
     * static method to retrieve a new `ItemAdapter`
     *
     * @return a new ItemAdapter
     */
    public static <Item extends IItem<? extends RecyclerView.ViewHolder>> ItemAdapter<Item> items() {
        return new ItemAdapter<>();
    }
}

File: library-core/src/main/java/com/mikepenz/fastadapter/adapters/ItemAdapter.kt
package com.mikepenz.fastadapter.adapters

import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.IItem
import com.mikepenz.fastadapter.IItemList
import com.mikepenz.fastadapter.utils.InterceptorUtil

/**
 * Kotlin type alias to simplify usage for an all accepting ItemAdapter
 */
typealias GenericItemAdapter = ItemAdapter<IItem<out RecyclerView.ViewHolder>>

/**
 * Created by mikepenz on 27.12.15.
 * A general ItemAdapter implementation based on the AbstractAdapter to speed up development for general items
 */
open class ItemAdapter<Item : IItem<out RecyclerView.ViewHolder>> : ModelAdapter<Item, Item> {

    constructor() : super(InterceptorUtil.DEFAULT as (element: Item) -> Item?)

    constructor(itemList: IItemList<Item>) : super(itemList, InterceptorUtil.DEFAULT as (element: Item) -> Item?)

    companion object {

        /**
         * static method to retrieve a new `ItemAdapter`
         *
         * @return a new ItemAdapter
         */
        @JvmStatic
        fun <Item : IItem<out RecyclerView.ViewHolder>> items(): ItemAdapter<Item> {
            return ItemAdapter()
        }
    }
}
